File: companion/ios/ScreenshotToolCompanion/Network/WebSocketClient.swift
import Foundation
import UIKit

/// State of a single desktop connection
struct DesktopConnection: Identifiable {
    let id: String // desktop name
    let host: String
    let port: Int
    var state: ConnectionState = .disconnected
    var webSocket: URLSessionWebSocketTask?
    var reconnectAttempts = 0
    var shouldReconnect = true
}

enum ConnectionState: Equatable {
    case disconnected
    case connecting
    case connected
}

/// Manages WebSocket connections to multiple desktop PixyVibe instances simultaneously.
class WebSocketClient: ObservableObject {
    static let shared = WebSocketClient()

    @Published var connections: [String: DesktopConnection] = [:]

    var connectedCount: Int { connections.values.filter { $0.state == .connected }.count }
    var hasAnyConnection: Bool { connectedCount > 0 }

    private var session: URLSession?
    private let deviceId = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
    private let maxReconnectAttempts = 5
    private var backgroundTask: UIBackgroundTaskIdentifier = .invalid

    private let appGroupDefaults = UserDefaults(suiteName: "group.com.pixyvibe.companion")

    /// User-configurable device name
    var deviceName: String {
        get {
            UserDefaults.standard.string(forKey: "custom_device_name")
                ?? UIDevice.current.name
        }
        set {
            UserDefaults.standard.set(newValue, forKey: "custom_device_name")
            appGroupDefaults?.set(newValue, forKey: "device_name_display")
            appGroupDefaults?.synchronize()
        }
    }

    private init() {
        session = URLSession(configuration: .default)
        setupBackgroundHandling()
    }

    // MARK: - Background Handling

    private func setupBackgroundHandling() {
        NotificationCenter.default.addObserver(
            forName: UIApplication.didEnterBackgroundNotification,
            object: nil, queue: .main
        ) { [weak self] _ in
            self?.beginBackgroundKeepAlive()
        }
        NotificationCenter.default.addObserver(
            forName: UIApplication.willEnterForegroundNotification,
            object: nil, queue: .main
        ) { [weak self] _ in
            self?.endBackgroundKeepAlive()
        }
    }

    private func beginBackgroundKeepAlive() {
        guard backgroundTask == .invalid else { return }
        backgroundTask = UIApplication.shared.beginBackgroundTask(withName: "KeepWebSocket") { [weak self] in
            self?.endBackgroundKeepAlive()
        }
    }

    private func endBackgroundKeepAlive() {
        guard backgroundTask != .invalid else { return }
        UIApplication.shared.endBackgroundTask(backgroundTask)
        backgroundTask = .invalid
    }

    // MARK: - Connection Management

    func connect(to desktop: DiscoveredDesktop) {
        guard let host = desktop.host, let port = desktop.port, desktop.isResolved else { return }
        connect(host: host, port: port, name: desktop.name)
    }

    func connect(host: String, port: Int, name: String) {
        // Skip if already connecting/connected to this desktop
        if let existing = connections[name], existing.state != .disconnected,
           existing.host == host, existing.port == port {
            return
        }

        // Disconnect existing connection to this desktop if any
        disconnectDesktop(name)

        let urlString = "ws://\(host):\(port)"
        guard let url = URL(string: urlString) else { return }

        let ws = session?.webSocketTask(with: url)
        ws?.resume()

        var conn = DesktopConnection(id: name, host: host, port: port, state: .connecting, webSocket: ws)
        conn.shouldReconnect = true
        connections[name] = conn

        DispatchQueue.main.async {
            self.objectWillChange.send()
        }

        // Save all connections for broadcast extension
        saveAllConnectionInfo()

        receiveMessages(for: name)
    }

    func disconnectDesktop(_ name: String) {
        guard var conn = connections[name] else { return }
        conn.shouldReconnect = false
        conn.webSocket?.cancel(with: .normalClosure, reason: nil)
        connections.removeValue(forKey: name)
        DispatchQueue.main.async {
            self.objectWillChange.send()
        }
        saveAllConnectionInfo()
    }

    func disconnectAll() {
        for name in connections.keys {
            connections[name]?.shouldReconnect = false
            connections[name]?.webSocket?.cancel(with: .normalClosure, reason: nil)
        }
        connections.removeAll()
        DispatchQueue.main.async {
            self.objectWillChange.send()
        }
        saveAllConnectionInfo()
    }

    /// Check if a desktop is connected or connecting
    func stateFor(_ desktopName: String) -> ConnectionState {
        connections[desktopName]?.state ?? .disconnected
    }

    // MARK: - App Group Storage

    private func saveAllConnectionInfo() {
        // Save list of desktops for broadcast extension
        let desktopList = connections.values.map { conn -> [String: Any] in
            ["host": conn.host, "port": conn.port, "name": conn.id]
        }

        if let data = try? JSONSerialization.data(withJSONObject: desktopList) {
            appGroupDefaults?.set(data, forKey: "desktop_list")
        }

        // Keep legacy single-desktop keys for backward compat
        if let first = connections.values.first {
            appGroupDefaults?.set(first.host, forKey: "desktop_host")
            appGroupDefaults?.set(first.port, forKey: "desktop_port")
            appGroupDefaults?.set(first.id, forKey: "desktop_name")
        }

        appGroupDefaults?.set(deviceId, forKey: "device_id")
        appGroupDefaults?.set(deviceName, forKey: "device_name_display")
        appGroupDefaults?.synchronize()
    }

    var lastDesktopName: String? {
        appGroupDefaults?.string(forKey: "desktop_name")
    }

    // MARK: - Sending

    /// Send a frame to all connected desktops
    func sendFrameToAll(_ jpegData: Data) {
        let base64 = jpegData.base64EncodedString()
        let dict: [String: Any] = ["type": "frame", "data": base64, "timestamp": Int(Date().timeIntervalSince1970)]
        for name in connections.keys {
            sendJSON(dict, to: name)
        }
    }

    /// Send a screenshot result to all connected desktops
    func sendScreenshotToAll(_ pngData: Data) {
        let base64 = pngData.base64EncodedString()
        let dict: [String: Any] = ["type": "screenshot_result", "data": base64]
        for name in connections.keys {
            sendJSON(dict, to: name)
        }
    }

    private func sendJSON(_ dict: [String: Any], to desktopName: String) {
        guard let conn = connections[desktopName],
              let json = try? JSONSerialization.data(withJSONObject: dict),
              let jsonString = String(data: json, encoding: .utf8)
        else { return }

        conn.webSocket?.send(.string(jsonString)) { error in
            if let error = error {
                print("[\(desktopName)] WebSocket send error: \(error)")
            }
        }
    }

    // MARK: - Receiving

    private func receiveMessages(for desktopName: String) {
        guard let conn = connections[desktopName] else { return }
        conn.webSocket?.receive { [weak self] result in
            guard let self = self else { return }

            switch result {
            case .success(let message):
                if case .string(let text) = message {
                    self.handleMessage(text, from: desktopName)
                }
                self.receiveMessages(for: desktopName)

            case .failure(let error):
                print("[\(desktopName)] WebSocket error: \(error)")
                DispatchQueue.main.async {
                    self.connections[desktopName]?.state = .disconnected
                    self.objectWillChange.send()
                }
                self.attemptReconnect(desktopName)
            }
        }
    }

    private func handleMessage(_ text: String, from desktopName: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String
        else { return }

        switch type {
        case "ping":
            let pong: [String: Any] = [
                "type": "pong",
                "device_name": deviceName,
                "device_id": deviceId
            ]
            sendJSON(pong, to: desktopName)

            DispatchQueue.main.async {
                self.connections[desktopName]?.state = .connected
                self.objectWillChange.send()
            }

        case "screenshot":
            print("[\(desktopName)] Screenshot request — triggering broadcast")
            BroadcastManager.shared.triggerBroadcast()

        default:
            break
        }
    }

    // MARK: - Reconnection

    private func attemptReconnect(_ desktopName: String) {
        guard var conn = connections[desktopName], conn.shouldReconnect else { return }

        guard conn.reconnectAttempts < maxReconnectAttempts else {
            print("[\(desktopName)] Max reconnect attempts reached")
            conn.shouldReconnect = false
            connections[desktopName] = conn
            DispatchQueue.main.async {
                self.connections.removeValue(forKey: desktopName)
                self.objectWillChange.send()
            }
            return
        }

        conn.reconnectAttempts += 1
        connections[desktopName] = conn
        let delay = min(pow(2.0, Double(conn.reconnectAttempts)), 30.0)
        let host = conn.host
        let port = conn.port

        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let self = self, self.connections[desktopName]?.shouldReconnect == true else { return }
            self.connect(host: host, port: port, name: desktopName)
        }
    }
}

File: companion/android/app/src/main/java/com/pixyvibe/companion/network/WebSocketClient.kt
package com.pixyvibe.companion.network

import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import com.pixyvibe.companion.discovery.DiscoveredDesktop
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import okhttp3.*
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.TimeUnit
import kotlin.math.min
import kotlin.math.pow

enum class ConnectionState {
    DISCONNECTED, CONNECTING, CONNECTED
}

data class DesktopConnection(
    val id: String,  // desktop name
    val host: String,
    val port: Int,
    val state: ConnectionState = ConnectionState.DISCONNECTED,
    val reconnectAttempts: Int = 0,
    val shouldReconnect: Boolean = true,
)

/**
 * Manages WebSocket connections to multiple desktop PixyVibe instances.
 * Mirrors iOS WebSocketClient.swift — same JSON protocol.
 */
class WebSocketClient(context: Context) {

    private val prefs: SharedPreferences =
        context.getSharedPreferences("pixyvibe_companion", Context.MODE_PRIVATE)

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _connections = MutableStateFlow<Map<String, DesktopConnection>>(emptyMap())
    val connections: StateFlow<Map<String, DesktopConnection>> = _connections.asStateFlow()

    val connectedCount: Int
        get() = _connections.value.values.count { it.state == ConnectionState.CONNECTED }

    val hasAnyConnection: Boolean
        get() = connectedCount > 0

    private val okClient = OkHttpClient.Builder()
        .pingInterval(30, TimeUnit.SECONDS)
        .build()

    private val webSockets = mutableMapOf<String, WebSocket>()

    val deviceId: String = prefs.getString("device_id", null)
        ?: UUID.randomUUID().toString().also { prefs.edit().putString("device_id", it).apply() }

    var deviceName: String
        get() = prefs.getString("custom_device_name", null) ?: Build.MODEL
        set(value) { prefs.edit().putString("custom_device_name", value).apply() }

    /** Callback invoked when desktop requests a screenshot. */
    var onScreenshotRequested: (() -> Unit)? = null

    /** Callback invoked when desktop requests start recording. */
    var onStartRecording: ((fps: Int) -> Unit)? = null

    /** Callback invoked when desktop requests stop recording. */
    var onStopRecording: (() -> Unit)? = null

    // MARK: - Connection Management

    fun connect(desktop: DiscoveredDesktop) {
        val host = desktop.host ?: return
        val port = desktop.port ?: return
        if (!desktop.isResolved) return
        connect(host, port, desktop.name)
    }

    fun connect(host: String, port: Int, name: String) {
        val existing = _connections.value[name]
        if (existing != null && existing.state != ConnectionState.DISCONNECTED
            && existing.host == host && existing.port == port) {
            return
        }

        // Disconnect existing if any
        disconnectDesktop(name)

        val conn = DesktopConnection(
            id = name, host = host, port = port,
            state = ConnectionState.CONNECTING,
            shouldReconnect = true,
        )
        updateConnection(name, conn)
        saveAllConnectionInfo()

        val request = Request.Builder()
            .url("ws://$host:$port")
            .build()

        val ws = okClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                updateConnection(name, _connections.value[name]?.copy(
                    state = ConnectionState.CONNECTED,
                    reconnectAttempts = 0,
                ))
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text, name)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                updateConnection(name, _connections.value[name]?.copy(
                    state = ConnectionState.DISCONNECTED,
                ))
                attemptReconnect(name)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                updateConnection(name, _connections.value[name]?.copy(
                    state = ConnectionState.DISCONNECTED,
                ))
                attemptReconnect(name)
            }
        })

        webSockets[name] = ws
    }

    fun disconnectDesktop(name: String) {
        val conn = _connections.value[name] ?: return
        updateConnection(name, conn.copy(shouldReconnect = false))
        webSockets.remove(name)?.close(1000, "User disconnect")
        val updated = _connections.value.toMutableMap()
        updated.remove(name)
        _connections.value = updated
        saveAllConnectionInfo()
    }

    fun disconnectAll() {
        for (name in _connections.value.keys.toList()) {
            updateConnection(name, _connections.value[name]?.copy(shouldReconnect = false))
            webSockets.remove(name)?.close(1000, "Disconnect all")
        }
        _connections.value = emptyMap()
        saveAllConnectionInfo()
    }

    fun stateFor(desktopName: String): ConnectionState {
        return _connections.value[desktopName]?.state ?: ConnectionState.DISCONNECTED
    }

    // MARK: - Sending

    fun sendFrameToAll(base64Jpeg: String) {
        val json = JSONObject().apply {
            put("type", "frame")
            put("data", base64Jpeg)
            put("timestamp", System.currentTimeMillis() / 1000)
        }.toString()

        for ((name, ws) in webSockets) {
            if (_connections.value[name]?.state == ConnectionState.CONNECTED) {
                ws.send(json)
            }
        }
    }

    fun sendScreenshotToAll(base64Png: String) {
        val json = JSONObject().apply {
            put("type", "screenshot_result")
            put("data", base64Png)
        }.toString()

        for ((name, ws) in webSockets) {
            if (_connections.value[name]?.state == ConnectionState.CONNECTED) {
                ws.send(json)
            }
        }
    }

    // MARK: - Message Handling

    private fun handleMessage(text: String, desktopName: String) {
        val json = try { JSONObject(text) } catch (_: Exception) { return }
        val type = json.optString("type")

        when (type) {
            "ping" -> {
                val pong = JSONObject().apply {
                    put("type", "pong")
                    put("device_name", deviceName)
                    put("device_id", deviceId)
                }.toString()
                webSockets[desktopName]?.send(pong)

                updateConnection(desktopName, _connections.value[desktopName]?.copy(
                    state = ConnectionState.CONNECTED,
                ))
            }
            "screenshot" -> {
                onScreenshotRequested?.invoke()
            }
            "start_recording" -> {
                val fps = json.optInt("fps", 10)
                onStartRecording?.invoke(fps)
            }
            "stop_recording" -> {
                onStopRecording?.invoke()
            }
        }
    }

    // MARK: - Reconnection

    private fun attemptReconnect(desktopName: String) {
        val conn = _connections.value[desktopName] ?: return
        if (!conn.shouldReconnect) return

        val maxAttempts = 5
        if (conn.reconnectAttempts >= maxAttempts) {
            val updated = _connections.value.toMutableMap()
            updated.remove(desktopName)
            _connections.value = updated
            return
        }

        val newAttempts = conn.reconnectAttempts + 1
        updateConnection(desktopName, conn.copy(reconnectAttempts = newAttempts))

        val delay = min(2.0.pow(newAttempts.toDouble()), 30.0).toLong() * 1000

        scope.launch {
            delay(delay)
            val current = _connections.value[desktopName]
            if (current?.shouldReconnect == true) {
                connect(current.host, current.port, desktopName)
            }
        }
    }

    // MARK: - Persistence

    private fun saveAllConnectionInfo() {
        val desktopList = JSONArray()
        for (conn in _connections.value.values) {
            desktopList.put(JSONObject().apply {
                put("host", conn.host)
                put("port", conn.port)
                put("name", conn.id)
            })
        }

        prefs.edit()
            .putString("desktop_list", desktopList.toString())
            .putString("device_id", deviceId)
            .putString("device_name_display", deviceName)
            .apply()
    }

    /** Get the saved desktop list (used by ScreenCaptureService). */
    fun getSavedDesktops(): List<Triple<String, String, Int>> {
        val raw = prefs.getString("desktop_list", null) ?: return emptyList()
        return try {
            val arr = JSONArray(raw)
            (0 until arr.length()).mapNotNull { i ->
                val obj = arr.getJSONObject(i)
                val name = obj.optString("name", "Desktop")
                val host = obj.optString("host") ?: return@mapNotNull null
                val port = obj.optInt("port", 0)
                if (port > 0) Triple(name, host, port) else null
            }
        } catch (_: Exception) {
            emptyList()
        }
    }

    private fun updateConnection(name: String, conn: DesktopConnection?) {
        if (conn == null) return
        val updated = _connections.value.toMutableMap()
        updated[name] = conn
        _connections.value = updated
    }

    fun destroy() {
        disconnectAll()
        scope.cancel()
        okClient.dispatcher.executorService.shutdown()
    }
}
